### ISSUE TITLE
New-TemporaryFile not available when calling powershell.exe from PowerShell 7 through cmd.exe or CreateProcess

### ISSUE THREAD

**Issue by hero101111** [PowerShell/PowerShell#14100]
## Steps to reproduce

```powershell
PS C:\test> cmd.exe /c "powershell.exe -command ""New-TemporaryFile"""
```

## Expected behavior

```none
Command to success and return a temporary file path
```

## Actual behavior

```none
Error of "The term 'New-TemporaryFile' is not recognized as the name of a cmdlet, function, script file..."
```

## Environment data

```none
PowerShell 7.1 on Windows 10 x64 version 2004 (for pwsh.exe)
Windows PowerShell 5.1 (for powershell.exe)
```

## Note
When calling`powershell
PS C:\test> powershell.exe -command "New-TemporaryFile"
`
the command works as expected. It seems that when passing through `cmd.exe` or `::CreateProcess()` the `Microsoft.PowerShell.Utility` assembly somehow is loaded differently and a few commands, including `New-TemporaryFile` end up missing.

New-TemproaryFile is available both on PowerShell 7 and on Windows PowerShell 5.1 so this is quite an interesting bug.

**Comment by vexx32** [comment]
That _is_ pretty odd! I don't think there's much, if anything, that PS 7 can do to rectify the behaviour you're seeing in Windows PowerShell.

 WinPS is not being developed any further, so I don't think it's likely that whatever is causing this in Windows PowerShell will be fixed.

**Comment by mklement0** [comment]
It's a moot point, as @vexx32 states, but the bug is related to `New-TemporaryFile` being implemented as a _function_ in Windows PowerShell.

In other words: it seems that all commands returned by `Get-Command -Module Microsoft.PowerShell.Utility -Type function` are affected in Windows PowerShell, and potentially all other commands implemented as _functions_.

In PowerShell Core, all former `Microsoft.PowerShell.Utility` _functions_ have been re-implemented as _cmdlets_, so the problem by definition doesn't arise there.

That said, in PowerShell Core there still are _functions_ among the in-box modules, as 
`Get-Command -Type function | where Source -like 'Microsoft.*'` indicates.

Calling such a function via the PowerShell _Core_  CLI seems to work fine, fortunately:

```
pwsh -noprofile -c 'Compress-Archive -?'
```

**Comment by iSazonov** [comment]
I guess it is again PSModulePath issue. If we run _directly_ Windows PowerShell we modify PSModulePath as needed, otherwise we don't.

**Comment by mklement0** [comment]
Good call, @iSazonov: PSModulePath in conjunction with _indirect_ invocation of `powershell.exe` is indeed the problem, given that the following works:

```
cmd.exe /c set PSModulePath=C:\WINDOWS\system32\WindowsPowerShell\v1.0\Modules `& powershell.exe -nop -c New-TemporaryFile
```

**Comment by mklement0** [comment]
I got myself confused earlier: I actually only see the problem in `7.1.0-rc.2`, and no longer in `7.1.0`.

Also, unless the following test is flawed, the problem does _not_ appear to be related to `$env:PSModulePath`, because I get equivalent results in both versions:

```
PSCore> cmd /c powershell.exe -nop -command `$env:PSModulePath -split "';'"
```

**Comment by hero101111** [comment]
@mklement0 I think you were on the right track before. When I run your command on `7.1.0` with the `cmd.exe` proxy I get `%ProgramFiles%\PowerShell\Modules` and `%ProgramFiles$\PowerShell\7\Modules` on top of the paths returned _without_ going through the `cmd.exe` proxy.

Looks like they have precedence over the others, when loading  `Microsoft.PowerShell.Utility`.

When running your variant, with manually setting `PSModulePath`, I get no errors, `New-TemporaryFile` is called successfully.

**Comment by mklement0** [comment]
Interesting, @hero101111; your experience is consistent with @iSazonov's explanation, but the strange thing is that the despite the presence of these extra, PS Core-only folders, the command succeeds for me in `7.1.0`.

**Comment by hero101111** [comment]
I just tried installing PowerShell 7.1.0 in `$env:USERPROFILE` and it still fails to locate `New-TemporaryFile`. Tried it also on a second Windows 10 version 2004 (Windows PowerShell 5.1.19041) machine as well. Seems like there's a different factor at play here.

**Comment by GaetanCambier** [comment]
Workaround : 
cmd.exe /c "powershell.exe -command ""Import-Module Microsoft.PowerShell.Utility;New-TemporaryFile"""

**Comment by gzm55** [comment]
another workaround

```powershell
cmd /c "set PSModulePath=&&powershell -command ""New-TemporaryFile"""
```

**Comment by aetonsi** [comment]
Just to add to the discussion, i personally resorted to calling the internal .NET function in case of error. Importing `Microsoft.PowerShell.Utility` or setting `PSModulePath=` didn't work for me.

```powershell
    try {
        $TempFile = New-TemporaryFile
    }
    catch {
        $TempFile = Get-Item ([System.IO.Path]::GetTempFilename())
    }
```

**Comment by RobMahn** [comment]
The problem is that Powershell V7 (pwsh.exe) adds "c:\program files\powershell\7\Modules" to PSModulePath and then Powershell V5 (powershell.exe) inherits the modified PSModulePath .  I don't know the reason that New-TemporaryFile has a problem with that when other commands do not.
The best solution is to call pwsh.exe instead of powershell.exe.  Or, if you're not sure what version of powershell is making the call, use $((Get-Process -Id $PID).Processname) .

**Comment by mklement0** [comment]
Thanks, @RobMahn; I just remembered something:

Yes, with _indirect_ execution - such as via `cmd /c` - the PS Core-only `$env:PSModulePath` entries are _retained_. While I don't know why `New-TemporaryFile` is specifically affected, it's not hard to see that problems can arise in this scenario. 

It is only with _direct_ execution that they are filtered out.

Thus, given that there's rarely a reason to call `powershell.exe` via `cmd /c` (or, for that matter, the PowerShell CLI from a PowerShell session altogether), the simpler workaround is to simply call `powershell.exe -command New-TemporaryFile` (or `pwsh -command New-TemporaryFile` or `cmd /c 'pwsh -command New-TemporaryFile'`)

**Comment by aetonsi** [comment]
@PowerShellTeam @SteveL-MSFT Please re-open issue as it was not solved
